---
title: "Trackviewer_olaps"
author: "Joel Moro"
date: "9/12/2021"
output: html_document
---

# 0. Libs
```{r message=FALSE, warning=FALSE}
library(GenomicRanges)
library(Gviz)
library(trackViewer)
library(rtracklayer)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(org.Hs.eg.db)
library(tidyr)
library(stringr)
library(Cairo)
```

# 1. Set Up

```{r}
#Specify Population, species and genes.
pop <- c("Tet", "TFH", "TH0")
species <- "mouse"
plots <- c("Il10", "Ccr5", "Maf", "Tox2", "Tcf7", "Lef1", "Irf4")
#plots <- c("Il10", "Ccr5")

roundUpNice <- function(x, nice=c(1,2,4,5,6,8,10)) {
  if(length(x) != 1) stop("'x' must be of length 1")
  10^floor(log10(x)) * nice[[which(x <= 10^floor(log10(x)) * nice)[[1]]]]
}
average <- function(A, B) {(A+B)/2}
average3 <- function (A,B) {(A+B)/3}


if (species == "mouse"){
  TxDb <- TxDb.Mmusculus.UCSC.mm10.knownGene
  org.SYMBOL2EG <- org.Mm.egSYMBOL2EG
  org.SYMBOL <- org.Mm.egSYMBOL
} else {
  TxDb <- TxDb.Hsapiens.UCSC.hg38.knownGene
  org.SYMBOL2EG <- org.Hs.egSYMBOL2EG
  org.SYMBOL <- org.Hs.egSYMBOL
}
```


```{r}
#Import results files from the different analysis to be included and for each replicate:
#List all the files in you data directory and import them depending on their format
#IMPORTANT: name files always as type_of_sample(Tet, Tconv, TFH...)+number of replicate_tecnique (ATAC, RNA...), e.g. Tet3_ATAC, Tet1_RNA...

file_list_bed <- list.files(path= paste0(getwd(), "/data/trackviewer/bed"),
                              pattern= "*.bed")
  
file_list_bw <- list.files(path= paste0(getwd(), "/data/trackviewer/bw"),
                             pattern= "*.bw")

assays <- c("ATAC", "RNA", "k4", "k27", "K27me3", "enh", "methy")

#Set up color vector (SAME LENGTH AS ASSAYS PLEASE!!!)
mycolors <- c("#E69F00", "#56B4E9", "#009E73",
          "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```


# 2. Loop
```{r}
for (xy in plots) {
  id <- get(xy, org.SYMBOL2EG)
  gr <- genes(TxDb)[id]
  agr <- if (width(gr)<99000) {
    resize(gr, 100000, fix = "center") 
  } else {
    resize(gr, width(gr)+20000, fix = "center")
  }
  genesinrange <- mapRangesToIds(TxDb, agr, type = "gene", ignore.strand = T)
  tracks <- sapply(genesinrange[[1]][[1]], 
                   function(z) {
                     track <- geneTrack(z,TxDb)[[1]]
                     return(track)
                   })
  
  for (file in c(file_list_bed)){
    x <- importScore(paste0("data/trackviewer/bed/", file), 
                     format = "BED",
                     ignore.strand = T,
                     ranges = agr)
    assign(paste(strsplit(file, ".", fixed=T)[1][[1]][1]), x)
  }
  
  for (file in c(file_list_bw)){
    x <- importScore(paste0("data/trackviewer/bw/", file), 
                     format = "BigWig",
                     ignore.strand = T,
                     ranges = agr)
    assign(paste(strsplit(file, ".", fixed=T)[1][[1]][1]), x)
  }
  
  #comp <- c("Ctl", "Smp")
  comp <- pop
  
  # Calculate average peaks for all assays and samples
  for (l in 1:length(pop)){
    for (s in assays){  
      for (i in 1:length(grep(paste0(pop[l], ".*" , s), names(.GlobalEnv), value=TRUE))){
        me <- eval(as.symbol(grep(paste0(pop[l], ".*" , s), names(.GlobalEnv),value=TRUE)[i]))
        assign(paste0(s, ".", comp[l], i), me)
      }
    }
    #indicate experiments with multiple replicates
    for (n in c("ATAC", "RNA")){
      #assign all objects to be used for operations
      for(w in 1:length(grep(paste0(n, "*.",comp[l], "."), names(.GlobalEnv),value=TRUE))){
        a <- eval(as.symbol(grep(paste0(n, ".",comp[l], "."), names(.GlobalEnv),value=TRUE)[w]))
        assign(paste0("mit",w), a)
      }
      
      if (length(grep("mit", names(.GlobalEnv),value=TRUE)) == 2) {
        avr <- mit1
        avr$dat <- avr$dat
        avr$dat2 <- mit2$dat
        avr$dat <- GRoperator(avr$dat, avr$dat2, 
                              col="score", operator=average)
        avr$dat2 <- GRanges()
        assign(paste0("arch", n, ".", comp[l]), avr)
      }
      
      if (length(grep("mit", names(.GlobalEnv),value=TRUE)) == 3){
        avr <- mit1
        avr$dat <- avr$dat
        avr$dat2 <- mit2$dat
        avr$dat <- GRoperator(avr$dat, avr$dat2, 
                              col="score", operator="+")
        avr$dat2 <- mit3$dat
        avr$dat <- GRoperator(avr$dat, avr$dat2, 
                              col="score", operator=average3)
        avr$dat2 <- GRanges()
        assign(paste0("arch", n, ".", comp[l]), avr)
      }
      
      if (length(grep("mit", names(.GlobalEnv),value=TRUE)) == 4){
        avrg <- mit1
        avrg$dat <- avrg$dat
        avrg$dat2 <- mit2$dat
        avrg$dat <- GRoperator(avrg$dat, avrg$dat2, 
                               col="score", operator=average)
        avrg$dat2 <- GRanges()
        
        avrg2 <- mit3
        avrg2$dat <- avrg2$dat
        avrg2$dat2 <- mit4$dat
        avrg2$dat <- GRoperator(avrg2$dat, avrg2$dat2, 
                                col="score", operator=average)
        avrg2$dat2 <- GRanges()
        
        avr <- avrg
        avr$dat <- avr$dat
        avr$dat2 <- avrg2$dat
        avr$dat <- GRoperator(avr$dat, avr$dat2, 
                              col="score", operator=average)
        avr$dat2 <- GRanges()
        assign(paste0("arch", n, ".", comp[l]), avr)
      }
      rm(list=grep("mit", names(.GlobalEnv), value = T))
    }
  }
  
  
  # Create tracks for each group, add "arch" for replicates:
  show.tracks <- c(mget(grep("methy.",names(.GlobalEnv),value=TRUE, fixed = T)),
                   mget(grep("enh.",names(.GlobalEnv),value=TRUE, fixed = T)),
                   mget(grep("k4.",names(.GlobalEnv),value=TRUE, fixed = T)),
                   mget(grep("K27me3.",names(.GlobalEnv),value=TRUE, fixed = T)),
                   mget(grep("k27.",names(.GlobalEnv),value=TRUE, fixed = T)),
                   mget(grep("archATAC.",names(.GlobalEnv),value=TRUE, fixed = T)),
                   mget(grep("archRNA.",names(.GlobalEnv),value=TRUE, fixed = T)))
  
  # Change the track names
  names(tracks) <- sapply(names(tracks), 
                          function(z) {
                            names <- get(z, org.Mm.egSYMBOL)
                            return(names)
                          })
  
  pdf(file = paste0("figs/",xy, ".pdf"), width = 10, height = 6)
  for (m in 1:length(pop)){
    Pattern_list<-do.call("list",mget(grep(comp[m],names(show.tracks),value=TRUE)))
    o <- optimizeStyle(trackList(Pattern_list, tracks))
    assign(paste0("optSty", m), o)
    u <- o$tracks
    assign(paste0("trackList", m), t)
    t <- o$style
    assign(paste0("viewerStyle", m), u)
    
    # Adjust X axis to show scale instead of chromosome ruler:
    setTrackViewerStyleParam(t, "xaxis", FALSE)
    setTrackViewerStyleParam(t, "margin", c(.01, .05, .01, .01))
    
    setTrackXscaleParam(u[[1]], "draw", FALSE)
    setTrackXscaleParam(u[[7]], "draw", TRUE) #REDUNDANT??
    setTrackXscaleParam(u[[7]], "gp", list(cex=.7))
    # Move scale in y axis to the right:
    setTrackViewerStyleParam(t, "margin", c(.08, .15, .01, .1))
    #Modify range in order to exclude the last names which include the gene names
    for(i in 1:length(assays)){
      setTrackYaxisParam(u[[i]], "main", FALSE)
    }
    for(i in c(1,3:length(assays))){
      setTrackYaxisParam(u[[i]], "gp", list(cex=.9))
    }
    # Eliminate scale for Active enhancers track:
    setTrackYaxisParam(u[[2]], "draw", FALSE)
    ## Adjust the limit of y-axis for RNA, ATAC and ChIP tracks:
    setTrackStyleParam(u[[1]], "ylim", c(0, 1))
    setTrackStyleParam(u[[2]], "ylim", c(0, 1))
    for (j in 3:length(assays)) {
    setTrackStyleParam(u[[j]], "ylim", c(0, roundUpNice(max(optimizeStyle(trackList(do.call("list",mget(grep(comp[1],names(show.tracks),value=TRUE))), tracks))$tracks[[j]]$dat$score,          optimizeStyle(trackList(do.call("list",mget(grep(comp[2],names(show.tracks),value=TRUE))), tracks))$tracks[[j]]$dat$score,
optimizeStyle(trackList(do.call("list",mget(grep(comp[3],names(show.tracks),value=TRUE))), tracks))$tracks[[j]]$dat$score))))
}

    for(i in 1:length(assays)){
      setTrackStyleParam(u[[i]], "marginBottom", .1)
    }
    # For each transcript, the transcript name can be put on the upstream or downstream 
    #of the transcript
    for(i in 8:length(u)){
      grx <- genes(TxDb)[as.data.frame(genesinrange)[i-length(assays),1]]
      if (start(ranges(grx)) <= start(ranges(agr))){
          setTrackStyleParam(u[[i]], "ylabpos", "left")
      } 
      if(end(ranges(grx)) >= end(ranges(agr))){
        setTrackStyleParam(u[[i]], "ylabpos", "right")
      } else {
        setTrackStyleParam(u[[i]], "ylabpos", "upstream")
      } 
      setTrackStyleParam(u[[i]], "ylabgp", list(cex=1))
    }
    # Adjust the label of y-axis (colour and size)
    for (i in 1:length(assays)){
      setTrackStyleParam(u[[i]], "ylabgp", list(cex=.8, col="black")) 
    }
    # Adjust track colors:
    for (i in 1:length(mycolors)) {
      setTrackStyleParam(u[[i]], "color", mycolors[i])
    }
    for(i in (length(assays)+1):length(u)){
      setTrackStyleParam(u[[i]], "color", "grey35")
    }
    
    names(u) <- c("DIFF METH", "DIFF ACT ENH", "K4", "K27me3", "K27",
                  "ATAC", "RNA", names(tracks))
    
    diff <- length(u) - length(assays)
    # Adjust the track height for assays:
    for (i in c(1:length(assays))){
      setTrackStyleParam(u[[i]], "height", (1-diff*0.04)/length(assays))
    }
    # Adjust track heigth
    for (i in (length(assays)+1):length(u)){
        setTrackStyleParam(u[[i]], "height", 0.04)
      }

    
    if (m == 1){
      vp <- viewTracks(u, gr=agr, viewerStyle=t, newpage = F)
    } else {
      vp <- viewTracks(u, gr=agr, viewerStyle=t, newpage = T)
    }
    addGuideLine(c(start(gr), end(gr)), vp=vp)
    grid.text(pop[m], x=0.07, y=0.95, just="top", 
              gp=gpar(cex=1.5, fontface="bold"))
  }
  dev.off()
}
?grid.text
```


```{r}

```


```{r}

```


```{r}

```

```{r}

```


```{r}

```

```{r}

```

```{r}

```


```{r}

```

```{r}

```

```{r}

```


```{r}

```


```{r}

```

```{r}

```

```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```


```{r}

```

```{r}

```

```{r}

```


```{r}

```


```{r}

```

```{r}

```


```{r}

```


```{r}

```

```{r}

```

```{r}

```


```{r}

```


```{r}

```


```{r}

```

```{r}

```

```{r}

```

```{r}

```

