# HiChIP analysis :thread:

This pipeline can be used to analyse HiChIP data. Briefly, HiChIP is a technique used to identify regions that are in contact in the 3D structure of the genome. In particular, we are connecting promoters to the regions they are contacting with by using a promoter enrichment procedure (chromatin immunoprecipitation) in the HiChIP protocol (see reference in /doc).  
</br>

## HiC-Pro

HiChIP data can be analysed using the `HiC-Pro` pipeline from Servant (for more information see the [documentation](http://nservant.github.io/HiC-Pro/QUICKSTART.html) or their [GitHub page](https://github.com/nservant/HiC-Pro)).

HiC-Pro was designed to process Hi-C data, from raw fastq files (paired-end Illumina data) to the normalized contact maps. HiC-Pro supports the main Hi-C protocols, including digestion protocols as well as protocols that do not require restriction enzyme such as DNase Hi-C. In practice, HiC-Pro can be used to process dilution Hi-C, in situ Hi-C, DNase Hi-C, Micro-C, capture-C, capture Hi-C or HiChIP data. The pipeline is flexible, scalable and optimized. HiC-Pro is sequential and each step of the workflow can be run independently.

HiC-Pro requires several dependencies; we are using Singurality to install the HiC-Pro pipeline and all its dependencies at once.  
</br>

In order to run HiC-Pro, you need to:

1) Prepare annotation files.

2) Put all your input files in a rawdata folder.

3) Set up your configuration file _config-install.txt_.

4) Run HiC-Pro.  
</br>

### Annotation files

In order to process the raw data, HiC-Pro requires 3 annotation files.

**1) A BED file of the restriction fragments after digestion.** \* The genome in the HiChIP protocol is fragmented into small pieces by using an enzymatic digestion. For example, we have used DpnII. See an example of this file right below:

````
chr1   0       16007   HIC_chr1_1    0   +
chr1   16007   24571   HIC_chr1_2    0   +
chr1   24571   27981   HIC_chr1_3    0   +
chr1   27981   30429   HIC_chr1_4    0   +
chr1   30429   32153   HIC_chr1_5    0   +
chr1   32153   32774   HIC_chr1_6    0   +
chr1   32774   37752   HIC_chr1_7    0   +
chr1   37752   38369   HIC_chr1_8    0   +
chr1   38369   38791   HIC_chr1_9    0   +
chr1   38791   39255   HIC_chr1_10   0   +
(...)
````

This file depends both on the restriction enzyme and the reference genome. It can be generated by 'in-silico' fragmenting the genome with the enzyme cutting sequence. We recommend generating this file in R by using the _Generate_annotation_files.R_ script. \* This file generates also the chromosome sizes file (see next section).

Alternatively, this can be done using the utility `digest_genome.py` from the HiC-Pro tool. For example, if you want to diggest the mm10 genome with DpnII, you need to have the genome in fasta (_mm10.fa_) and know the enzyme cutting sequence (you can extract that in the [NEB website](https://international.neb.com)). DpnII is cutting at GATC and you indicate the cutting point with an \^.

````
HICPRO_PATH/bin/utils/digest_genome.py -r ^GATC -o mm10_dpnii.bed mm10.fa
`````
</br>

**2) A table file of chromosomes’ size.**

````
chr1    249250621
chr2    243199373
chr3    198022430
chr4    191154276
chr5    180915260
chr6    171115067
chr7    159138663
chr8    146364022
chr9    141213431
chr10   135534747
(...)
````

You will generate this file when using the _Generate_annotation_files.R_ script mentioned above.

Alternatively, you can dowload this from the [UCSC page](http://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/) as a _mm10.chrom.sizes_ file.  
</br>

**3) The bowtie2 indexes.** See detailed information on how to get or generate Bowtie2 indexes in the [ATACseq info](https://github.com/patriciasolesanchez/PSlab/tree/master/ATACseq_analysis).  
</br>


### Raw data folder
Put all input files in a single data folder. The input files have to be organized with a folder per sample (i.e. create a folder per sample and add fastq files for each sample in the corresponding folder).  
</br>


### Configuration file
Copy and edit the configuration file _config-hicpro.txt_ in your local folder. See more details on how to edit the configuration file [here](https://nservant.github.io/HiC-Pro/MANUAL.html#manual).

The _config-hicpro.txt_ file that you can find here has our tested parameters to work with mouse HiChIP data. The most important points you need to check are:

1. PAIRX_EXT: check that the suffix from your paired reads corresponds to the one indicated in the file (e.g. _\_read1_, _\_read2_ or _\_R1_, _\_R2_).
2. BOWTIE2_IDX_PATH: indicate the folder where the genome indexes are found.
3. REFERENCE_GENOME: indicate the prefix your genome indexes have (This might be _bowtie_, not necessarily the genome name, e.g. mm10!)
4. GENOME_SIZE: indicate the full path to the chromosome sizes file.
5. GENONE_FRAGMENT: indicate the full path to the digested genome bed file.
6. LIGATION_SITE: indicate the sequence that is generated after ligating your digested genome. It depends on the cutting sequence of your restriction enzyme (e.g. DpnII cuts at GATC and the ligation site is GATCGATC).
</br>

### Run HiC-Pro

In order to use HiC-Pro, you can check tool options by using:

````
HiC-Pro --help
````

And then run HiC-Pro by typing:
````
MY_INSTALL_PATH/bin/HiC-Pro -i FULL_PATH_TO_RAW_DATA -o FULL_PATH_TO_OUTPUTS -c MY_LOCAL_CONFIG_FILE
````

**IMPORTANT!** You need to indicate an output folder that does not previously exist. If you run the pipeline again on an already existing folder, the pipeline will give error and not run.
</br>

You can run your HiC-Pro analysis in the cluster by directly submitting the _hicpro.sh_ script.

````
bsub < hicpro.sh
````
</br>


The HiC-Pro is run in the following steps:

#### 1) Reads Mapping
Each mate is independantly aligned on the reference genome. The mapping is performed in two steps. First, the reads are aligned using an end-to-end aligner. Second, reads spanning the ligation junction are trimmmed from their 3’ end, and aligned back on the genome. Aligned reads for both fragment mates are then paired in a single paired-end BAM file. Singletons and multi-hits can be discarded according the confirguration parameters. 
**IMPORTANT!** If the LIGATION_SITE parameter in the not defined, HiC-Pro will skip the second step of mapping! Alignment rate will be much worse.

#### 2) Fragment assignment and filtering
Each aligned read can be assigned to one restriction fragment according to the reference genome and the restriction enzyme. The next step is to separate the invalid ligation products from the valid pairs. Dangling end and self circles pairs are therefore excluded. Only **valid pairs** involving two different restriction fragments are used to build the contact maps. Duplicated valid pairs associated to PCR artefacts are discarded.

#### 3) Quality Controls
HiC-Pro performs QC for most of the analysis steps. The alignment statistics are the first QC. Aligned reads in the first (end-to-end) step, and alignment after trimming are reported (in pratice, we ususally observed around 10-20% of trimmed reads). _\* An abnormal level of trimmed reads can reflect a ligation issue._  
Once the reads are aligned on the genome, HiC-pro checks the number of singleton, multiple hits or duplicates. The fraction of valid pairs are presented for each type of ligation products. You would expect a balance in valid pairs ligation type (1/4 each).  
Finally HiC-Pro also calculates the distribution of fragment size on a subset of valid pairs. Additional statistics will report the fraction of intra/inter-chromosomal contacts, as well as the proportion of short range (<20kb) versus long range (>20kb) contacts.

#### 4) Map builder
Intra and inter-chromosomal contact maps are build for all specified resolutions. The genome is splitted into bins of equal size. Each valid interaction is associated with the genomic bins to generate the raw maps.

#### 5) ICE normalization
HiC-Pro uses a fast implementation of the original ICE normalization algorithm (Imakaev et al. 2012), making the assumption of equal visibility of each fragment.  
</br>

### RESULTS
All outputs follow the input organization, with one folder per sample. You can find:

#### bowtie_results
This folder contains the results of the reads mapping. The results of 1st mapping step are available in the _bwt2_glob_ folder, and the 2nd step in the _bwt2_loc_ folder. Final BAM files, reads pairing, and mapping statistics are available on the _bwt2_ folder. **IMPORTANT!** Once HiC-Pro has been run, all files in _bwt2_glob_ or _bwt2_loc_ folders can be removed.

#### hic_results
This folder contains all Hi-C processed data, and is further divided in several sub-folders.

* Data: stores the valid interaction products (.validPairs), as well as other statisics files. One validPairs file is generated per reads chunck. These files are then merged in the allValidPairs, and duplicates are removed if specified in the configuration file. The _validPairs_ are stored using a simple tab-delimited text format:
````
read name / chr_reads1 / pos_reads1 / strand_reads1 / chr_reads2 / pos_reads2 / strand_reads2 / fragment_size [/ allele_specific_tag]
````

* Matrix: contains the contact maps. This folder is organized with raw and iced contact maps for all resolutions. Only no zero values are stored. BED files describing the genomic bins are also generated. Note that abs and ord files are identical in the context of Hi-C data as the contact maps are symmetric. Contact maps are stored as a triplet sparse format:
````
bin_i / bin_j / counts_ij
````

* Pic: contains graphical outputs of the quality control checks.  
</br>


NEXT STEPS: peak calling and differential analysis (FitHiChIP and HiChIP_peaks tools).


## Contact Matrix (Juicer Tools)

Juicebox is visualization software for Hi-C data. In specific, we will take advantage of their gui to visualize the contact matrix.
Firstly, it is necessary to convert our .ValidPairs obtained with HiC-Pro into a .hic that can be used as input for juicebox. (scripts are all available in the cluster software/ directory.

**Call Format**:
````
/PATH/TO/hicpro2juicebox.sh -i /PATH/TO/.allValidPairs \
    -g /PATH/TO/chrom_hg19.sizes \
    -j juicer_tools.jar \
    -o  OUTDIR/
````

**Example Call**: 
````
/gpfs/projects/cek26/software/scripts/hicpro2juicebox.sh -i /gpfs/projects/cek26/experiments/HiChip/outs27_04/hic_results/data/dixon_2M_2/dixon_2M_2.allValidPairs \
    -g /gpfs/projects/cek26/experiments/HiChip/data/chrom_hg19.sizes \
    -j /gpfs/projects/cek26/software/juicer_tools_1.22.01.jar \
    -o  /gpfs/projects/cek26/experiments/HiChip/juicebox_out
````

As a result, it will retrieve the .hic files that can be later introduced into the juicebox GUI.
Juicebox gui can be downloaded in: https://github.com/aidenlab/Juicebox/wiki/Download. With simple GUIs for Windows and Mac and .jar for UNIX.

**Result Example**:
![alt text](https://raw.githubusercontent.com/JGarnica22/PSlab/c4668269e6f0269a5e9db883151c7cfa340662a0/HiChIP_analysis/doc/2022.04.27.14.01.17.HiCImage.svg?token=AO2KRHY57NM2RBSBMRZCTRLCNEYUQ)

## Loop Calling (FitHiChip)

## Differential Analysis
