# HiChIP analysis :thread:

This pipeline can be used to analyse HiChIP data. Briefly, HiChIP is a technique used to identify regions that are in contact in the 3D structure of the genome. In particular, we are connecting promoters to the regions they are contacting with by using a promoter enrichment procedure (chromatin immunoprecipitation) in the HiChIP protocol (see reference in /doc).  
</br>

## HiC-Pro

HiChIP data can be analysed using the `HiC-Pro` pipeline from Servant (for more information see the [documentation](http://nservant.github.io/HiC-Pro/QUICKSTART.html) or their [GitHub page](https://github.com/nservant/HiC-Pro)).

HiC-Pro was designed to process Hi-C data, from raw fastq files (paired-end Illumina data) to the normalized contact maps. HiC-Pro supports the main Hi-C protocols, including digestion protocols as well as protocols that do not require restriction enzyme such as DNase Hi-C. In practice, HiC-Pro can be used to process dilution Hi-C, in situ Hi-C, DNase Hi-C, Micro-C, capture-C, capture Hi-C or HiChIP data. The pipeline is flexible, scalable and optimized. HiC-Pro is sequential and each step of the workflow can be run independently.

HiC-Pro requires several dependencies; we are using Singurality to install the HiC-Pro pipeline and all its dependencies at once.  
</br>

In order to run HiC-Pro, you need to:

1) Prepare annotation files.

2) Put all your input files in a rawdata folder.

3) Set up your configuration file _config-install.txt_.

4) Run HiC-Pro.  
</br>

### Annotation files

In order to process the raw data, HiC-Pro requires 3 annotation files.

**1) A BED file of the restriction fragments after digestion.** \*The genome in the HiChIP protocol is fragmented into small pieces by using an enzymatic digestion. For example, we have used DpnII. See an example of this file right below:

````
chr1   0       16007   HIC_chr1_1    0   +
chr1   16007   24571   HIC_chr1_2    0   +
chr1   24571   27981   HIC_chr1_3    0   +
chr1   27981   30429   HIC_chr1_4    0   +
chr1   30429   32153   HIC_chr1_5    0   +
chr1   32153   32774   HIC_chr1_6    0   +
chr1   32774   37752   HIC_chr1_7    0   +
chr1   37752   38369   HIC_chr1_8    0   +
chr1   38369   38791   HIC_chr1_9    0   +
chr1   38791   39255   HIC_chr1_10   0   +
(...)
````

This file depends both on the restriction enzyme and the reference genome. It can be generated by 'in-silico' fragmenting the genome with the enzyme cutting sequence. We recommend generating this file in R by using the _Generate_annotation_files.R_ script.

Alternatively, this can be done using the utility `digest_genome.py`. For example, if you want to diggest the mm10 genome with DpnII, you need to have the genome in fasta (_mm10.fa_) and know the enzyme cutting sequence (you can extract that in the [NEB website](https://international.neb.com)). DpnII is cutting at GATC and you indicate the cutting point with an \^.

````
HICPRO_PATH/bin/utils/digest_genome.py -r ^GATC -o mm10_dpnii.bed mm10.fa
`````
</br>

</br>

**2) A table file of chromosomesâ€™ size.**

````
chr1    249250621
chr2    243199373
chr3    198022430
chr4    191154276
chr5    180915260
chr6    171115067
chr7    159138663
chr8    146364022
chr9    141213431
chr10   135534747
(...)
````

You can dowload this from the [UCSC page](http://hgdownload.soe.ucsc.edu/goldenPath/mm10/bigZips/) as a _mm10.chrom.sizes_ file.  
</br>

**3) The bowtie2 indexes.** See the bowtie2 manual page for details about how to create such indexes.


### Raw data folder


### Configuration file


### Run HiC-Pro

````
MY_INSTALL_PATH/bin/HiC-Pro -i FULL_PATH_TO_RAW_DATA -o FULL_PATH_TO_OUTPUTS -c MY_LOCAL_CONFIG_FILE
````

You can run your HiC-Pro analysis in the cluster by directly submitting the _hicpro.sh_ script.
