---
title: "DE_analysis_RNAseq_Notebook"
author: "Josep Garnica"
date: "16/03/2020"
output:
  pdf_document: default
  html_document: default
---
This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. This script is to perform DE analysis with RNAseq data. It was created for R 3.6.2 version (2019-12-12).Copyright (C) 2020  Patricia Sole Sanchez.

#Differential expression analysis of RNAseq data
```{r raw data treatment and PCA, include=F, warning=FALSE, message=FALSE}
# Check if required packages are installed, if not install:,
cran.packages <- c("ggplot2", "ggrepel", "pheatmap", "RColorBrewer")
for (i in cran.packages) {
  if(!require(i, character.only = TRUE)) {
    install.packages(i)
    print(paste(i,"just installed"))
  }
  else {
    print(paste(i,"was already installed"))
  }
}
bioc.packages <- c("DESeq2", "biomaRt")
for (i in bioc.packages) {
  if (!require(i, character.only = TRUE)) {
    BiocManager::install(i)
    print(paste(i,"just installed"))
  } else {
    print(paste(i,"was already installed"))
  }
}
# Load packages:
library(ggplot2)
library(ggrepel)
library(pheatmap)
library(RColorBrewer)
library(DESeq2)
library(biomaRt)
# Set your working directory (the project you are working in):
setwd("C:/Users/jgarn/OneDrive - Universitat de Barcelona/Documentos/Bioinformatics/Functionals/DE_analysis_RNAseq")
# Specify parameters to be used along the script:
## Indicate name of txt file containing expression data (raw counts)
expfile <- "Partek_TFH_Raw_counts.txt"
## Indicate populations of interest, to be compared:
# First indicate control population, then sample:
pop <- c("Th0", "TFH")
# Load data:
counts <- read.table(paste0("data/", expfile),
                     header = T,
                     sep = "\t",
                     quote = "", 
                     dec = ".")
# Set gene_name as rows labels
row.names(counts) <- counts$gene_name
# Eliminate unnecessary columns, we only want gene_name and counts for the different populations.
counts2 <- counts[, c(grep(pop[1], colnames(counts)), grep(pop[2], colnames(counts)))]
# Change names of columns to shorter, easier ones:
names (counts2) <- c(paste0(rep(pop[1], length(grep(pop[1],colnames(counts2)))),
                            1:length(grep(pop[1],colnames(counts2)))), 
                     paste0(rep(pop[2], length(grep(pop[2],colnames(counts2)))),
                            1:length(grep(pop[2],colnames(counts2)))))
# Create conditions:
sample_info <- data.frame(cell_type = c(rep(pop[1], length(grep(pop[1], colnames(counts2)))), 
                                        rep(pop[2], length(grep(pop[2], colnames(counts2))))),
                          replicates = c(1:length(grep(pop[1],colnames(counts2))),
                                         1:length(grep(pop[2],colnames(counts2)))),
                          row.names = names (counts2))
# Transform the matrix by rounding the counts and transforming the values from "numeric" to "integer"
counts2_r <- apply(counts2, c(1,2), round)
counts2_i <- apply(counts2_r, c(1,2), as.integer)
# Generate the DESeq2 dataset:
dds <- DESeqDataSetFromMatrix (countData = counts2_i,
                               colData = sample_info,
                               design = ~ cell_type)
# Remove genes with very low counts:
dds <- dds[ rowSums (counts(dds)) > 10, ]
#DESeqâs default method to normalize read counts to account for differences in 
#sequencing depths is implemented in estimateSizeFactors
DESeq.ds_f <- estimateSizeFactors(dds)
sizeFactors(DESeq.ds_f)
#With rlog() function from DESeq2, we directly obtain counts normalized for sequencing 
#depth and transformed to log2 scale
DESeq.rlog <- rlog(DESeq.ds_f, blind = TRUE)
rlog.norm.counts <- assay (DESeq.rlog)
write.table(rlog.norm.counts, file = "output/rlog_normalized_DESeq.txt", quote = FALSE, sep = "\t", dec = ".")

# Plot PCA with prcomp():
#prcomp calculates principal components
pc <- prcomp(t(rlog.norm.counts))
pc_sum <- summary(pc)
PC1_varexpl <- pc_sum$importance[2,"PC1"]
PC2_varexpl <- pc_sum$importance[2,"PC2"]

pdf(file = "figs/PCA.pdf", width = 5, height = 5)
plot(x = pc$x[,1], y= pc$x[,2],
     col = c("red", "blue")[colData(DESeq.ds_f)[,1]],
     pch = 16,
     xlab = paste0("PC1 = ", round(PC1_varexpl*100,1), "%"),
     ylab = paste0("PC2 = ", round(PC2_varexpl*100,1), "%"),
     las = 1,
     main = "PCA of seq. depth normalized \n and rlog - transformed read counts ")
dev.off()
```
##PCA
```{r PCA, echo=F, warning=FALSE, message=FALSE}
plot(x = pc$x[,1], y= pc$x[,2],
     col = c("red", "blue")[colData(DESeq.ds_f)[,1]],
     pch = 16,
     xlab = paste0("PC1 = ", round(PC1_varexpl*100,1), "%"),
     ylab = paste0("PC2 = ", round(PC2_varexpl*100,1), "%"),
     las = 1,
     main = "PCA of seq. depth normalized \n and rlog - transformed read counts ")
```
```{r Prepare vol, include=FALSE, warning=FALSE, message=FALSE}
# Set control population for the DE comparison:
dds$cell_type <- relevel(dds$cell_type, pop[1])
# Apply DESeq to do differential expression:
dds <- DESeq(dds)
# Show dataframe from differential expression analysis. If you want, change
# the direction of comparison
res <- results(dds, contrast=c("cell_type", pop[2], pop[1]))
# Clean up genes with low counts and high variability with Shrinkage
resLFC <- lfcShrink(dds, coef=resultsNames(dds)[2], type="apeglm")
# Order genes based on their pvalue
resOrdered <- resLFC[order(resLFC$log2FoldChange, decreasing = T),]
# Volcano plo:
Vol <- data.frame(FC= resOrdered$log2FoldChange,
                  sig= -log10(resOrdered$padj))
plot <- ggplot(Vol, aes(x=FC, y=sig))

pdf(file = "figs/Volcano_plot.pdf", width = 5, height = 5)
plot + geom_point( aes(color=sig>2 & abs(FC)>2),
                   size = 0.3,
                   show.legend = F) + 
  scale_color_manual(values = c("TRUE" = "red","FALSE"="black")) +
  geom_vline(xintercept = c(-2, 2), linetype = "dashed") +
  geom_hline(yintercept = 2, linetype = "dashed") +
  #MODIFY IF NEEDED:
  #scale_y_continuous(breaks = c(0, 5, 10, 50, 100, 200, 250), 
  #limits =  c(0, 250), trans = "sqrt") +
  #xlim(-10, 10) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        aspect.ratio = 0.9) +
  ggtitle(paste0(strsplit(resultsNames(dds)[2], "_")[[1]][3], " vs. ",
                 strsplit(resultsNames(dds)[2], "_")[[1]][5])) +
  xlab("log2FC") + ylab("-log10(padj)")
dev.off()
```
##Volcano plot
```{r Volcano plot, echo=F, warning=FALSE, message=FALSE}
print(paste("Differential expressed genes between",paste(strsplit(resultsNames(dds)[2], "_")[[1]][3], "and",
                 strsplit(resultsNames(dds)[2], "_")[[1]][5])),)
plot + geom_point( aes(color=sig>2 & abs(FC)>2),
                   size = 0.3,
                   show.legend = F) + 
  scale_color_manual(values = c("TRUE" = "red","FALSE"="black")) +
  geom_vline(xintercept = c(-2, 2), linetype = "dashed") +
  geom_hline(yintercept = 2, linetype = "dashed") +
  #MODIFY IF NEEDED:
  #scale_y_continuous(breaks = c(0, 5, 10, 50, 100, 200, 250), 
  #limits =  c(0, 250), trans = "sqrt") +
  #xlim(-10, 10) +
  theme_bw() +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        aspect.ratio = 0.9) +
  ggtitle(paste0(strsplit(resultsNames(dds)[2], "_")[[1]][3], " vs. ",
                 strsplit(resultsNames(dds)[2], "_")[[1]][5])) +
  xlab("log2FC") + ylab("-log10(padj)")
```
```{r Biomart mouse prep, include=F, warning=FALSE, message=FALSE}
# BIOMART: add gene types and ensembl IDs to DESeq2 output
# Select Mart:
ensembl <- useMart("ENSEMBL_MART_ENSEMBL")
# Select a Dataset:
ensembl <- useDataset("mmusculus_gene_ensembl", mart=ensembl)
# If you need to look per specific attribute, filter or dataset:
#searchAttributes(mart = ensembl, pattern = "ensembl.*id")
# Find in Biomart ensembleID, gene type and mouse gene name:
BM <- getBM(attributes=c("ensembl_gene_id","gene_biotype",
                          "external_gene_name"), mart = ensembl)
# Now merge this data frame with the DESeq2 results:
DesBM <- merge(x= as.data.frame(resOrdered), y=BM, 
               by.x="row.names", by.y = "external_gene_name", 
               all.x = T)
colnames(DesBM)[1] <- paste("Gene name")
write.table(DesBM, paste0("output/DESeq2_", 
                          strsplit(resultsNames(dds)[2], "_")[[1]][3], "_vs_",
                          strsplit(resultsNames(dds)[2], "_")[[1]][5], ".txt"), 
            quote = F, dec = ".", sep = "\t")
DESgenes <- DesBM[which(abs(DesBM$log2FoldChange)>10 & DesBM$padj<0.0001),]
write.table(DESgenes, paste0("output/DESeq2_FC10_padj0.0001_", 
                          strsplit(resultsNames(dds)[2], "_")[[1]][3], "_vs_",
                          strsplit(resultsNames(dds)[2], "_")[[1]][5], ".txt"), 
            quote = F, dec = ".", sep = "\t")
```
##Heatmap
```{r heatmap most differential, echo=F, warning=FALSE, message=FALSE}
print(paste("Differential expression in genes with log2Foldchange>10 and padjusted<0<0001 between",
            strsplit(resultsNames(dds)[2], "_")[[1]][3], "and",
                 strsplit(resultsNames(dds)[2], "_")[[1]][5]))
library(knitr)
kable(DESgenes[,c(1,3,4,6)], digits = 15, row.names = F, )

# HEATMAP:
DEgenes <- DesBM[which(abs(DesBM$log2FoldChange)>10 & DesBM$padj<0.0001), "Row.names"]

pheatmap(as.matrix(rlog.norm.counts[DEgenes,]), scale = "row", 
         rev(brewer.pal(8, ("PiYG"))),
         # cluster_rows = F, cluster_cols = F, na_col = "white",
         # border_color = "black",
          cellwidth = 20, cellheight = 10,
         # main = "Table 1", 
         # fontsize = 10,
         # fontsize_row = 9,
         # fontsize_col = 12, angle_col = 90,
         # width = 800, height = 4000)
)

```
```{r echo=F, warning=FALSE, message=FALSE}
library(rmarkdown)
library(yaml)
render("DE_analysis_RNAseq_Notebook.Rmd", output_format ="html_document")
```

